### ISSUE TITLE
Python 3.9 AttributeError: __args__

### ISSUE THREAD

**Issue by GameDungeon** [Instagram/MonkeyType#231]
I honestly cannot parse this error message.

I'm trying to run monkeytype on manim's mobject.py

```
monkeytype stub manim.mobject.mobject

Traceback (most recent call last):
  File "C:\Python39\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Python39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\ethan\AppData\Local\pypoetry\Cache\virtualenvs\manim-ob3XvX2_-py3.9\Scripts\monkeytype.exe\__main__.py", line 7, in <module>
  File "c:\users\ethan\appdata\local\pypoetry\cache\virtualenvs\manim-ob3xvx2_-py3.9\lib\site-packages\monkeytype\cli.py", line 396, in entry_point_main
    sys.exit(main(sys.argv[1:], sys.stdout, sys.stderr))
  File "c:\users\ethan\appdata\local\pypoetry\cache\virtualenvs\manim-ob3xvx2_-py3.9\lib\site-packages\monkeytype\cli.py", line 381, in main
    handler(args, stdout, stderr)
  File "c:\users\ethan\appdata\local\pypoetry\cache\virtualenvs\manim-ob3xvx2_-py3.9\lib\site-packages\monkeytype\cli.py", line 199, in print_stub_handler
    stub = get_stub(args, stdout, stderr)
  File "c:\users\ethan\appdata\local\pypoetry\cache\virtualenvs\manim-ob3xvx2_-py3.9\lib\site-packages\monkeytype\cli.py", line 126, in get_stub
    stubs = build_module_stubs_from_traces(
  File "c:\users\ethan\appdata\local\pypoetry\cache\virtualenvs\manim-ob3xvx2_-py3.9\lib\site-packages\monkeytype\stubs.py", line 819, in build_module_stubs_from_traces
    return build_module_stubs(defns)
  File "c:\users\ethan\appdata\local\pypoetry\cache\virtualenvs\manim-ob3xvx2_-py3.9\lib\site-packages\monkeytype\stubs.py", line 784, in build_module_stubs
    imports = get_imports_for_signature(entry.signature)
  File "c:\users\ethan\appdata\local\pypoetry\cache\virtualenvs\manim-ob3xvx2_-py3.9\lib\site-packages\monkeytype\stubs.py", line 157, in get_imports_for_signature
    param_imports = get_imports_for_annotation(param.annotation)
  File "c:\users\ethan\appdata\local\pypoetry\cache\virtualenvs\manim-ob3xvx2_-py3.9\lib\site-packages\monkeytype\stubs.py", line 143, in get_imports_for_annotation
    elem_types = anno.__args__ or []
  File "C:\Python39\lib\typing.py", line 694, in __getattr__
    raise AttributeError(attr)
AttributeError: __args__
```

**Comment by GameDungeon** [comment]
Version 21.5.0 for MonkeyType

**Comment by GameDungeon** [comment]
@carljm This is becoming quite an issue for me as it chooses seemingly random files that I am completely unable to stub with monkeytype. Can you offer any advice?

**Comment by GameDungeon** [comment]
this seems quite similar to #122 but has a different error. I think the only difference is the versions though.

**Comment by carljm** [comment]
@GameDungeon There is some kind of type present in your annotations that passes the `is_generic()` test at https://github.com/Instagram/MonkeyType/blob/master/monkeytype/stubs.py#L137 but yet doesn't have an `__args__` attribute, which most generic types should have, and so fails at https://github.com/Instagram/MonkeyType/blob/master/monkeytype/stubs.py#L143

It would be useful to know what type this is; probably once we know it won't be hard to add support for it. If you are able to temporarily add a `try: ... except AttributeError: import pdb; pdb.set_trace()` around line 143 locally, it will drop you into pdb (the Python debugger) when the error is hit, which should allow you to explore what value the `anno` name references. If you have trouble with this, if you can provide a precise set of instructions to reproduce the problem using publicly-available code, I will look into it when I can.

**Comment by GameDungeon** [comment]
The error comes up when trying to stub [camera.py](https://github.com/ManimCommunity/manim/blob/master/manim/camera/camera.py) and [mobject.py](https://github.com/ManimCommunity/manim/blob/master/manim/mobject/mobject.py) both are untyped as of now. I'll try the except.

**Comment by GameDungeon** [comment]
Ok, I think I found it: WARNING: Failed decoding trace: Attribute specified by 'Callable' in module 'typing' is of type <class 'typing._CallableType'>, not type. When opening the terminal in admin it works but tells me traces failed to decode. Verbose output gives me the above.

**Comment by GameDungeon** [comment]
Also if you see this later, quick question. Does MonkeyType allow for custom generics?

**Comment by carljm** [comment]
@GameDungeon The `Callable` trace decoding failure doesn't seem clearly related to the `__args__` AttributeError.

If you want me to repro I'll need more detailed instructions; in particular the key missing step is how you are collecting the traces in the first place.

MonkeyType doesn't record generic arg types for custom generics; it has no way to know how to introspect an instance of an arbitrary generic type to find the right types for its type args.

**Comment by GameDungeon** [comment]
I'm currently calling monkeytype run on this. 
```python
import inspect
import importlib
import manim
import os
import sys


def all_tests() -> None:
    sys.path.append('tests')
    all_files = list()

    for root, dirs, files in os.walk("."):
        if "__pycache__" in root:
            continue
        for file in files:
            all_files.append(file)
        for directory in dirs:
            if directory != "__pycache__":
                sys.path.append(f"{root}\\{directory}")

    for file in all_files:
        if file.startswith("test_"):
            try:
                importlib.import_module(file)
            except Exception as e:
                print(f"Can't run this test: {file}, Beacuse: {e}")

    print("Tests Done")

all_tests()

```

It's the only way I could call monkey type run and get it to work through our tests.

**Comment by GameDungeon** [comment]
I think this error might be the missing piece in the puzzle:
```
Failed to serialize trace
Traceback (most recent call last):
  File "c:\python39\lib\site-packages\monkeytype\encoding.py", line 249, in serialize_traces
    yield CallTraceRow.from_trace(trace)
  File "c:\python39\lib\site-packages\monkeytype\encoding.py", line 211, in from_trace
    return_type = maybe_encode_type(type_to_json, trace.return_type)
  File "c:\python39\lib\site-packages\monkeytype\encoding.py", line 174, in maybe_encode_type
    return encode(typ)
  File "c:\python39\lib\site-packages\monkeytype\encoding.py", line 146, in type_to_json
    type_dict = type_to_dict(typ)
  File "c:\python39\lib\site-packages\monkeytype\encoding.py", line 82, in type_to_dict
    qualname = typ.__qualname__
  File "c:\python39\lib\typing.py", line 694, in __getattr__
    raise AttributeError(attr)
AttributeError: __qualname__
```
I get a ton of those when running MonkeyType run on the above. Need anything else @carljm?

**Comment by GameDungeon** [comment]
I have to belive the above error is directly causing:

> WARNING: Failed decoding trace: Attribute specified by 'Callable' in module 'typing' is of type <class 'typing._CallableType'>, not type

Still see no clear connection between that and the original error.

**Comment by carljm** [comment]
I'm having trouble even getting types collected. After a bunch of work to install all the manim dependencies, I'm now getting this error:

```
➜ monkeytype run collect.py
Manim Community v0.6.0

[05/27/21 11:04:24] ERROR    LaTeX compilation error! LaTeX reports:                                                                                          tex_file_writing.py:180
Traceback (most recent call last):
  File "collect.py", line 70, in run_function
    obj(**end_args)
TypeError: __init__() missing 1 required positional argument: 'scale_to_resolution'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/carljm/Library/Caches/pypoetry/virtualenvs/manim-RhTJCDRG-py3.8/bin/monkeytype", line 33, in <module>
    sys.exit(load_entry_point('MonkeyType', 'console_scripts', 'monkeytype')())
  File "/Users/carljm/projects/monkeytype/monkeytype/cli.py", line 396, in entry_point_main
    sys.exit(main(sys.argv[1:], sys.stdout, sys.stderr))
  File "/Users/carljm/projects/monkeytype/monkeytype/cli.py", line 381, in main
    handler(args, stdout, stderr)
  File "/Users/carljm/projects/monkeytype/monkeytype/cli.py", line 224, in run_handler
    runpy.run_path(args.script_path, run_name='__main__')
  File "/Users/carljm/.pythonz/pythons/CPython-3.8.7/lib/python3.8/runpy.py", line 265, in run_path
    return _run_module_code(code, init_globals, run_name,
  File "/Users/carljm/.pythonz/pythons/CPython-3.8.7/lib/python3.8/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/Users/carljm/.pythonz/pythons/CPython-3.8.7/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "collect.py", line 129, in <module>
    all_classes()
  File "collect.py", line 125, in all_classes
    run_function(obj, dict())
  File "collect.py", line 75, in run_function
    arg_gen = arg_generator()
  File "collect.py", line 39, in __init__
    "matrix": Matrix([[2, 2], [2, 2]]),
  File "/Users/carljm/projects/manim/manim/manim/mobject/matrix.py", line 130, in __init__
    mob_matrix = self.matrix_to_mob_matrix(matrix)
  File "/Users/carljm/projects/manim/manim/manim/mobject/matrix.py", line 144, in matrix_to_mob_matrix
    return np.vectorize(self.element_to_mobject)(
  File "/Users/carljm/Library/Caches/pypoetry/virtualenvs/manim-RhTJCDRG-py3.8/lib/python3.8/site-packages/numpy/lib/function_base.py", line 2113, in __call__
    return self._vectorize_call(func=func, args=vargs)
  File "/Users/carljm/Library/Caches/pypoetry/virtualenvs/manim-RhTJCDRG-py3.8/lib/python3.8/site-packages/numpy/lib/function_base.py", line 2191, in _vectorize_call
    ufunc, otypes = self._get_ufunc_and_otypes(func=func, args=args)
  File "/Users/carljm/Library/Caches/pypoetry/virtualenvs/manim-RhTJCDRG-py3.8/lib/python3.8/site-packages/numpy/lib/function_base.py", line 2151, in _get_ufunc_and_otypes
    outputs = func(*inputs)
  File "/Users/carljm/projects/manim/manim/manim/mobject/svg/tex_mobject.py", line 432, in __init__
    SingleStringMathTex.__init__(
  File "/Users/carljm/projects/manim/manim/manim/mobject/svg/tex_mobject.py", line 259, in __init__
    file_name = tex_to_svg_file(
  File "/Users/carljm/projects/manim/manim/manim/utils/tex_file_writing.py", line 44, in tex_to_svg_file
    dvi_file = compile_tex(
  File "/Users/carljm/projects/manim/manim/manim/utils/tex_file_writing.py", line 202, in compile_tex
    while not tex_lines[environment - 1].startswith("\\end"):
NameError: name 'tex_lines' is not defined
```

**Comment by GameDungeon** [comment]
How did you install Latex?

**Comment by GameDungeon** [comment]
That is strange. It seems you got the github version too. I'll see if I can replicate the error.

**Comment by carljm** [comment]
FWIW I'm on OS X and installed latex via `brew install basictex`. FWIW it looks like that `tex_lines` variable really is not defined, so I'm guessing you must be somehow avoiding that codepath altogether.

**Comment by GameDungeon** [comment]
Is running every function with the right parameters a healthy way to use monkeytype?

**Comment by carljm** [comment]
> Is running every function with the right parameters a healthy way to use monkeytype?

It's not an approach I've seen before. It should work, but the types you get out will be only as good as what you put in. Typically you'd run MonkeyType with a more realistic workload (test suite and/or some kind of "production" workload), and for stuff you can't get covered that way, you'd add types a different way. E.g. if you can reliably predict the types of args based on their name, then you could write a LibCST transform (even based on MonkeyType's) to just add those types to args with those names if not already annotated, without going through MonkeyType for it.

**Comment by GameDungeon** [comment]
"but the types you get out will be only as good as what you put in" If I put in a bad type and it errors will monkey type still log it?

**Comment by carljm** [comment]
Yeah if a function raises an exception, MonkeyType will still log the argument types, just no return type for that particular execution.

**Comment by GameDungeon** [comment]
While I believe that any function that raises an exception shouldn't be logged, I'll keep that in mind. I went this route as the test suite for manim is lacking.

**Comment by carljm** [comment]
> While I believe that any function that raises an exception shouldn't be logged

There are plenty of cases where functions raise exceptions that are not related to a bad type of argument being passed in. There could even be cases where a certain type of argument always results in an exception, but this is part of the contract of the function and that argument type should still be annotated as valid for that argument!

MonkeyType's job is to record what happens at runtime and tell you about the types it saw, not to make judgements (which will inevitably not be right for all cases) about which executions were valid or invalid.

**Comment by GameDungeon** [comment]
Ok I see your point, I'll just have to be careful then. 
I ran it again and got instead of the normal error:
WARNING: Failed decoding trace: Attribute specified by 'Type' in module 'typing' is of type <class 'typing._SpecialGenericAlias'>, not type.
WARNING: Failed decoding trace: Module 'builtins' has no attribute 'list_iterator'

**Comment by GameDungeon** [comment]
@carljm Soon I'm just going to move on from this issue, as it has taken a lot of time, but let me know if you need my help.

**Comment by GameDungeon** [comment]
Now I'm getting this lovely error: AttributeError: '_SpecialForm' object has no attribute '__name__'

**Comment by ntjess** [comment]
I reproduced the __args__ error with a much easier setup:

```python
# monkeymodule/file1.py
def to_str(func):
  def wrapper(*args):
    newArgs = tuple(map(str, args))
    return func(*newArgs)
  return wrapper

@to_str
def add(a, b):
  return a + b
```

```python
# monkeymodule/file2.py
from file1 import add
add(1,2)
```

Terminal script:
```bash
monkeytype.exe run .\monkeymodule\file2.py
monkeytype.exe stub monkeymodule.file1
```

Trace:
```
Traceback (most recent call last):
  File "<userpath>\miniconda3\envs\py39\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "<userpath>\miniconda3\envs\py39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "<userpath>\Miniconda3\envs\py39\Scripts\monkeytype.exe\__main__.py", line 7, in <module>
  File "<userpath>\miniconda3\envs\py39\lib\site-packages\monkeytype\cli.py", line 396, in entry_point_main
    sys.exit(main(sys.argv[1:], sys.stdout, sys.stderr))
  File "<userpath>\miniconda3\envs\py39\lib\site-packages\monkeytype\cli.py", line 381, in main
    handler(args, stdout, stderr)
  File "<userpath>\miniconda3\envs\py39\lib\site-packages\monkeytype\cli.py", line 199, in print_stub_handler
    stub = get_stub(args, stdout, stderr)
  File "<userpath>\miniconda3\envs\py39\lib\site-packages\monkeytype\cli.py", line 126, in get_stub
    stubs = build_module_stubs_from_traces(
  File "<userpath>\miniconda3\envs\py39\lib\site-packages\monkeytype\stubs.py", line 819, in build_module_stubs_from_traces
    return build_module_stubs(defns)
  File "<userpath>\miniconda3\envs\py39\lib\site-packages\monkeytype\stubs.py", line 784, in build_module_stubs
    imports = get_imports_for_signature(entry.signature)
  File "<userpath>\miniconda3\envs\py39\lib\site-packages\monkeytype\stubs.py", line 157, in get_imports_for_signature
    param_imports = get_imports_for_annotation(param.annotation)
  File "<userpath>\miniconda3\envs\py39\lib\site-packages\monkeytype\stubs.py", line 143, in get_imports_for_annotation
    elem_types = anno.__args__ or []
  File "<userpath>\miniconda3\envs\py39\lib\typing.py", line 694, in __getattr__
    raise AttributeError(attr)
AttributeError: __args__
```

**Comment by GameDungeon** [comment]
ok, it seems to break on decorators? still wondering on the "AttributeError: '_SpecialForm' object has no attribute '__name__'"

**Comment by carljm** [comment]
Thanks @ntjess for the simple repro! It looks like in Python 3.9 unbound `typing.Callable` has no `__args__` attribute, whereas in previous Python versions it had `__args__` of empty tuple. Should be a pretty easy fix.

**Comment by carljm** [comment]
@GameDungeon 

> Now I'm getting this lovely error: AttributeError: '_SpecialForm' object has no attribute '__name__'. This has been a fun time.

Sorry for the unpleasant experience. MonkeyType uses internal API of the `typing` module that changes on every Python release, and the 3.9 support is not very well exercised yet. If you have the option of using Python 3.8 instead, I expect you might have a better experience. But thanks for helping flush out bugs in the 3.9 support!

If you want to file a separate issue with a full traceback for the `SpecialForm has no attribute __name__` I can see if I can repro it.

**Comment by carljm** [comment]
Fixed with merge of #237.
